lib/bolt/transport/ssh/exec_connection.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+require 'open3'
+
+module Bolt
+  module Transport
+    class SSH < Simple
+      class ExecConnection
+        attr_reader :user, :target
+
+        def initialize(target)
+          @target = target
+          ssh_config = Net::SSH::Config.for(target.host)
+          @user = @target.user || ssh_config[:user] || Etc.getlogin
+          @logger = Logging.logger[self]
+          validate(target)
+        end
+
+        def connect
+          _, err, stat = Open3.capture3(build_ssh_command('exit'))
+          unless stat.success?
+            raise Bolt::Node::ConnectError.new(
+              "Failed to connect to #{@target.safe_name}: #{err}",
+              'CONNECT_ERROR'
+            )
+          end
+        end
+
+        def shell
+          Bolt::Shell::Bash.new(@target, self)
+        end
+
+        def userhost
+          "#{@user}@#{@target.host}"
+        end
+
+        def build_ssh_command(command)
+          ssh_conf = @target.transport_config['ssh-command']
+          ssh_cmd = if ssh_conf.is_a?(String)
+                      [ssh_conf]
+                    elsif ssh_conf.is_a?(Array)
+                      ssh_conf
+                    end
+          ssh_cmd += %W[-p #{@target.port}] if @target.port
+          ssh_cmd += %w[-o BatchMode=yes]
+          ssh_cmd += %w[-o StrictHostKeyChecking=no] if @target.transport_config['host-key-check'] == false
+          if (key = target.transport_config['private-key'])
+            if key.instance_of?(String)
+              ssh_cmd += %W[-i #{key}]
+            else
+              # This is simpler than writing out a tempfile for now
+              raise Bolt::ValidationError, 'private-key must be a filepath when using ssh-command'
+            end
+          end
+          ssh_cmd << userhost
+          ssh_cmd << command
+          Shellwords.shelljoin(ssh_cmd)
+        end
+
+        def copy_file(source, dest)
+          @logger.debug { "Uploading #{source}, to #{userhost}:#{dest}" } unless source.is_a?(StringIO)
+          cp_cmd = @target.transport_config['copy-command'] || ["scp", "-r"]
+          cp_cmd += %W[-P #{@target.port}] if @target.port
+          cp_cmd += %w[-o StrictHostKeyChecking=no] if @target.transport_config['host-key-check'] == false
+          if (key = target.transport_config['private-key'])
+            if key.instance_of?(String)
+              cp_cmd += %W[-i #{key}]
+            else
+              # This is simpler than writing out a tempfile for now
+              raise Bolt::ValidationError, 'private-key must be a filepath when using ssh-command'
+            end
+          end
+
+          _, err, stat = if source.is_a?(StringIO)
+                           Tempfile.create(File.basename(dest)) do |f|
+                             f.write(source.read)
+                             f.close
+                             cp_cmd << f
+                             cp_cmd << "#{userhost}:\"#{dest}\""
```

Comment:
I think we need to shellescape `dest` rather than quoting it. Apparently that argument is interpreted by the shell on the target before the file is written. Who knew?

Reply:
I don't think just shellescape will work. With that, scp sees `scp /local/file with spaces user@host:file with spaces`. It correctly interprets the first argument as '/local/file with spaces', but needs to have quotes around the destination otherwise it interprets the spaces as new arguments. It looks like there are other escapes we can do, but quotes seems just as reasonable as other options: https://stackoverflow.com/questions/19858176/how-do-i-escape-spaces-in-path-for-scp-copy-in-linux

Reply:
Ah yes I see, you're not quoting it, you're escaped-quoting it. Then the options are either to double shellescape it (technically shellescape `dest` and then shellescape the entire argument) or to shellescape it once and use the array form of open3 rather than string.